docs/tutorial/perms.md

```diff
@@ -0,0 +1,195 @@
+
```

Comment:
Cette nouvelle page n'a pas un ton de documentation, mais ressemble plutôt aux explications d'un mec salé par un système complexe qu'il s'occupe à juger très fort.
Elle fait disparaitre des choses importantes comme `le système de permissions intégré dans Django ne suffisait pas`, qui était au tout début de l'ancienne page. C'est quand même pas pour rien qu'on a fini avec ce système, crois moi que j'ai retourné le problème un sacré moment avec les permissions de Django avant de me lancer là dedans à l'époque. Je sais pas si aujourd'hui y'a des mécanismes différents qui aideraient, mais ce qui est sûr, c'est qu'en 2015, le système de permissions de Django ne permettait que de contrôler l'accès à des classes, et pas à des instances de classe, donc par exemple, c'était impossible d'avoir des visibilité différentes des messages de forums, des photos, ou des profiles utilisateurs, en fonction d'un paramètre de visibilité de chacun de ces objets.

Aussi, la section sur les perfomances est tout à fait bienvenue, parce que c'est très vrai que ça peut vite finir en désastre, mais encore une fois, le ton n'est pas du tout neutre, et c'est un problème. Une documentation n'est pas vraiment là pour porter des jugement sur l'esthétisme, qui est est purement subjectif, mais simplement pour expliquer d'un point de vue objectif les potentiels problèmes ou points forts, et globalement orienter le lecteur vers des réponses à "comment on s'en sert correctement?".

Reply:
> Cette nouvelle page n'a pas un ton de documentation, mais ressemble plutôt aux explications d'un mec salé par un système complexe qu'il s'occupe à juger très fort.

J'avoue que j'étais salé (et merci à Sli d'avoir un peu arrondi les bords).

> Elle fait disparaitre des choses importantes comme `le système de permissions intégré dans Django ne suffisait pas`, qui était au tout début de l'ancienne page. C'est quand même pas pour rien qu'on a fini avec ce système, crois moi que j'ai retourné le problème un sacré moment avec les permissions de Django avant de me lancer là dedans à l'époque. Je sais pas si aujourd'hui y'a des mécanismes différents qui aideraient, mais ce qui est sûr, c'est qu'en 2015, le système de permissions de Django ne permettait que de contrôler l'accès à des classes, et pas à des instances de classe, donc par exemple, c'était impossible d'avoir des visibilité différentes des messages de forums, des photos, ou des profiles utilisateurs, en fonction d'un paramètre de visibilité de chacun de ces objets.

Si, c'était possible : https://docs.djangoproject.com/en/1.8/ref/contrib/auth/#django.contrib.auth.models.User.has_perm
Alors certes, ça aurait quand même nécessité une surcouche de complexité, mais qui reste contrôlée par le cadre de Django et qui ne retire pas le reste du système de permissions.

Reply:
Non, ce qu'on a besoin n'était pas possible:
`has_perm` permet de passer un objet pour restreindre le jeux de permissions à vérifier pour un user, mais ne regarde systématique qu'une seule table de permission, donc par design ne peux pas implémenter un système de permissions en ManyToMany, où tu contrôles l'accès différemment pour différents objets du même type.

C'est vraiment comme essayer d'implémenter un ManyToMany sans table intermédiaire: tu peux toujours bricoler en remettant des listes dans les valeurs des tables, mais tu dois quand même résoudre le même problème à la fin, et tu perds l'optimisation des index de ta DB, parce que les listes sont opaques pour le moteur de DB, elle sont pas visibles depuis le schéma.  
En l'occurrence si tu joues avec les perms de Django, tu fais des perms du type `<app label>.<perm codename>.<object id>`, donc ta quantité de permission explose très vite très fort, et surtout, c'est très très difficile de maintenir la cohérence de tout le système sans avoir un script très complexe, et surtout très critique, quelque part qui tourne dans un cron pour remettre d'équerre le bouzin, ce qui n'était pas souhaitable à l'époque, parce que c'est un truc qui cassait régulièrement dans l'ancien site.

De plus, autant il faut faire globalement attention quand on code (mais ça c'est vrai partout, pas qu'avec les permissions), autant ça fait quand même bientôt 10 ans que les pages arrivent à être servies, et surtout, le système est suffisamment robuste et simple à apréhender pour que depuis le temps, y'aie pas eu à ma connaissance de gros soucis de permissions qui ont tout fait péter, du type escalade de privilège, malgré l'armée de bots qui a pris le truc d'assaut.

Reply:
Mais on n'est pas obligé de passer forcément par la table permission.
L'idée, c'est avant tout de rajouter un Backend qui ne vient pas remplacer le système de Django, mais le compléter. En rajoutant un backend, on en revient à faire ce qu'on veut pour le système de perms.
La doc de Django parle également de ce point : https://docs.djangoproject.com/fr/4.2/topics/auth/customizing/#handling-authorization-in-custom-backends

Alors certes, faire comme ça n'aurait pas forcément empêché l'explosion du nombre de requêtes pour voir une liste d'objets. Mais ça permet déjà d'être un peu plus dans les clous du framework, et donc de pouvoir en utiliser plus largement les fonctionnalités, sans pour autant renoncer à notre système.

Et bien entendu, il y a l'autre problème, celui de la modélisation des groupes, avec beaucoup d'entre-eux qui sont hardcodés, et certains qui marchent dans un sens, d'autre dans le sens inverse (les groupes de ban). C'est un système compliqué à utiliser et à maintenir.
En combinant le système de perms maison et le système de groupes maison, ça donne quelque chose avec lequel il n'est pas si évident de travailler.

> De plus, autant il faut faire globalement attention quand on code (mais ça c'est vrai partout, pas qu'avec les permissions)

Je suis vraiment pas d'accord avec cet argument. Ok il faut faire attention, mais ce n'est pas vrai partout à la même échelle. Simplement dire "il faut faire attention", ça n'empêche pas les développeur C/C++ de devoir gérer régulièrement des problèmes de segfault. Ce n'est pas une question de faire attention, mais une question d'architecture por circonscrire au minimum ce à quoi il faut faire attention.

Le système actuel pose des problèmes, en ce qu'il restreint ce qu'on peut faire, à cause des performances. Sur le forum, on devrait cacher tous les topics auquel l'utilisateur n'a pas le droit d'accéder. Mais on ne le fait pas. Et si on voulait le faire, on serait obligé de se poser la question des performances, alors que c'est quelque chose d'extrêmement basique.

Reply:
Le système de permission était déjà assez complexe pour que je ne vois pas trop l'intéret de se rajouter la contrainte de vouloir le greffer au framework, qui n'était quand même pas vraiment prévu pour ce qu'on avait besoin. Tu l'as dit toi même, ça n'aurait pas forcément empêché l'explosion des requêtes.
Pour les groupes, ça c'est aussi en partie historique: y'avait le même système de groupe dans l'ancien site (c'est calqué sur UNIX, qui n'est [pas si mal conçu](https://fr.wikipedia.org/wiki/Contr%C3%B4le_d%27acc%C3%A8s_%C3%A0_base_de_r%C3%B4les)), et la migration a permis que chacun retrouve les même droits directement, sans passer par une étape où il faut redonner les bonnes permissions à chacun, et le fonctionnement restait le même pour les administrateurs. Est-ce que c'est un bon choix, en rétrospective, je sais pas, mais je sais pas non plus comment se serait passé une migration manuelle et un réapprentissage complet d'un nouveau système de permission pour toute l'AE. Y'a eu déjà pas mal de friction lors de la transition.
Est-ce que c'est changeable? Probablement, mais à nouveau il faudrait faire face à la friction des utilisateurs (en plus du côté technique) qui devront réapprendre un système de permissions qui est pas tellement complexe de leur point de vue parce qu'il est assez intuitif (en tout cas c'était le cas avec mes utilisateurs de 2015: c'était quand même un sujet que j'avais discuté avec eux quelques fois).

Je ne reviens pas sur le "il faut faire attention", tu es passé à côté de l'argument qui était que "ça fait 10 ans que ça fonctionne en prod sans soucis majeur" :-)

Un système de permission est toujours un compromis entre fonctionnalités et performances. Y'a la même chose sur les systèmes d'exploitation: quand tu rajoutes SELinux ou autre [MAC](https://fr.wikipedia.org/wiki/Contr%C3%B4le_d%27acc%C3%A8s_obligatoire), tu perds forcément en performances, mais tu gagnes en fonctionnalités. Je ne suis pas sûr que si ça avait été basé sur le framework tu n'aurais pas le même soucis pour l'affichage du forum, puisqu'à la fin, le problème est le même: pour une page et un user, tu dois vérifier les permissions pour tous les objets à afficher, et ça sera forcément couteux, sauf à simplifier le système, mais donc on perd en granularité (c'est peut-être ce qu'il faut faire, après tout).

Je m'arrête là sur cette discussion, je t'ai pas mal refait l'historique et le pourquoi on en est là. Sûrement qu'on aurait pu faire autrement, mais est-ce que ça aurait été mieux? Sur tous les points? Sans autres compromis ailleurs? Et pis après tout, avec des "si"...
Hésite pas à ouvrir une issue en copiant collant les parties intéressantes discutées ici, si vraiment tu veux te lancer dans la réécriture des permissions, ça sera plus approprié.